File: examples/glife/Cell.cpp
#include "Cell.h"

using namespace std;

// Cellspace dimensions
long int Cell::w = 0;
long int Cell::h = 0;


Cell::Cell(long int x, long int y, long int w, long int h, Phase phase,
           short int nalive, Phase* vis_phase)
    : adevs::Atomic<CellEvent, int>(),
      x(x),
      y(y),
      phase(phase),
      nalive(nalive),
      vis_phase(vis_phase) {
    // Set the global cellspace dimensions
    Cell::w = w;
    Cell::h = h;
    // Set the initial visualization value
    if (vis_phase != NULL) {
        *vis_phase = phase;
    }
}

int Cell::ta() {
    // Check if a phase change should occur
    if (check_death_rule()     // cell will die
        || check_born_rule())  // cell will be birthed
    {
        return 1;
    }
    // Otherwise, do nothing
    return adevs_inf<int>();
}

void Cell::delta_int() {
    // One of these must be true
    assert(check_death_rule() || check_born_rule());
    // Change our state
    phase = (check_death_rule()) ? Dead : Alive;
}

void Cell::delta_ext(int e, list<CellEvent> const &xb) {
    // Update the count of living neighbors
    list<CellEvent>::const_iterator iter;
    for (iter = xb.begin(); iter != xb.end(); iter++) {
        if ((*iter).value == Dead) {
            nalive--;
        } else {
            nalive++;
        }
    }
}

void Cell::delta_conf(list<CellEvent> const &xb) {
    delta_int();
    delta_ext(0, xb);
}

void Cell::output_func(list<CellEvent> &yb) {
    CellEvent e;
    // Out is the state that we will assume
    e.value = (check_born_rule()) ? Alive : Dead;
    // Set the visualization value
    if (vis_phase != NULL) {
        *vis_phase = e.value;
    }
    // Generate an event for each neighbor
    for (long int dx = -1; dx <= 1; dx++) {
        for (long int dy = -1; dy <= 1; dy++) {
            e.x = (x + dx) % w;
            e.y = (y + dy) % h;
            if (e.x < 0) {
                e.x = w - 1;
            }
            if (e.y < 0) {
                e.y = h - 1;
            }
            // Don't send to self
            if (e.x != x || e.y != y) {
                yb.push_back(e);
            }
        }
    }
}

File: examples/java_life/Cell.java
import adevs.*;
import java.util.Collection;

/**
 * A single cell in the game of life.
 */
class Cell extends Atomic<Boolean>
{
	/**
	 * Create a cell at position x,y and set its initial
	 * state.
	 */
	public Cell(int x, int y, boolean is_alive)
	{
		super();
		this.x = x; 
		this.y = y;
		this.is_alive = is_alive;
		// Negative value triggers an output a t = 0
		alive_neighbors = -1;
	}
	/**
	 * Get the x position of the cell.
	 */
	public int getX() { return x; }
	/**
	 * Get the y position of the cell.
	 */
	public int getY() { return y; }
	/**
	 * Time advance function
	 */
	public double ta()
	{
		// Send initial output?
		if (alive_neighbors == -1 && is_alive)
			return 0.0;
		// If a phase change should occur
		else if (check_death_rule() // cell will die
			|| check_born_rule()) // cell will be birthed
			return 1.0;
		// Otherwise, do nothing
		return Double.MAX_VALUE;
	}
	/**
	 * Internal transition function
	 */
	public void delta_int() 
	{
		if (alive_neighbors < 0) alive_neighbors = 0;
		// Change the cell state if necessary
		if (check_death_rule()) is_alive = false;
		else if (check_born_rule()) is_alive = true;
	}
	/**
	 * External transition function
	 */
	public void delta_ext(double e, Collection<Boolean> xb) 
	{
		if (alive_neighbors < 0) alive_neighbors = 0;
		// Update the count if living neighbors
		for (Boolean x: xb)
		{
			if (x.booleanValue() == true)
				alive_neighbors++;
			else
				alive_neighbors--;
		}
	}
	/**
	 * Confluent transition function
	 */
	public void delta_conf(Collection<Boolean> xb) 
	{
		delta_int();
		delta_ext(0.0,xb);
	}
	/**
	 * Output function
	 */	
	public void output_func(Collection<Boolean> yb) 
	{
		// Check in case this in not true
		if (alive_neighbors == -1)
			yb.add(new Boolean(is_alive));
		else if (check_born_rule())
			yb.add(new Boolean(true));
		else yb.add(new Boolean(false));
	}
	/**
	 * Should the cell be birthed?
	 */
	private boolean check_born_rule()
	{
		return (!is_alive && alive_neighbors == 3);
	}
	/**
	 * Should the cell expire?
	 */
	private boolean check_death_rule()
	{
		return (is_alive && (alive_neighbors < 2 || alive_neighbors > 3));
	}

	private int x, y; // Position of the cell
	private int alive_neighbors; // Number of living neighbors
	private boolean is_alive; // Cell's state
}
